WT-7091 Restrict LSM tests with compatible backup configuration

raviprakashgiri29 opened the pull request.

kommiharibabu reviewed (approved):
LGTM

agorrod commented:
@raviprakashgiri29 would you mind updating our backup documentation to reflect this limitation? Bonus points for restricting the backup API usage if LSM trees are present and reporting a helpful error message.

raviprakashgiri29 commented:
> @raviprakashgiri29 would you mind updating our backup documentation to reflect this limitation? Bonus points for restricting the backup API usage if LSM trees are present and reporting a helpful error message.

Thanks, @agorrod for pointing this. Updated the documentation and added a condition to return an error when backup API is used with open LSM trees.

sueloverso commented:
These changes are fine for a work-around. But I'd like a new ticket to know what is happening for LSM because off-hand it should work for LSM. We should have a bug-fixing ticket to investigate that.

sueloverso commented on src/cursor/cur_backup.c:
```diff
@@ -333,6 +333,12 @@ __backup_add_id(WT_SESSION_IMPL *session, WT_CONFIG_ITEM *cval)
+    /* Return error if block-based incremental backup is performed with open LSM trees.  */
+    if (!TAILQ_EMPTY(&conn->lsmqh)) {
+        WT_RET_MSG(session, ENOTSUP, "LSM does not work with block-based incremental backup");
+    }
```
I think this should move (and use the `incremental_config` setting) to the compatibility testing chunk of code in `__backup_config` in this file.

raviprakashgiri29 replied:
Moved to check the condition to  `__backup_config ` function.

sueloverso commented on src/cursor/cur_backup.c:
```diff
@@ -333,6 +333,12 @@ __backup_add_id(WT_SESSION_IMPL *session, WT_CONFIG_ITEM *cval)
     conn = S2C(session);
     blk = NULL;
+
+    /* Return error if block-based incremental backup is performed with open LSM trees.  */
```
Minor: Return an error...

raviprakashgiri29 replied:
Fixed

sueloverso commented on src/docs/backup.dox:
```diff
@@ -171,6 +171,10 @@ database directory has not been opened and recovery run.  Once recovery
 
+The only limitation of block-based incremental backup is that it does not
+work with LSM trees. It will throw an error if backup API is used to do
+block-based incremental backup with LSM trees.
```
I suggest making this more of an active voice:
Block-based incremental backup does not work with LSM trees. An error will be returned in that case.

raviprakashgiri29 replied:
Updated

sueloverso reviewed (commented):
I have one request for moving the library code and a couple minor things. Otherwise this looks good and is very close.

raviprakashgiri29 commented:
> These changes are fine for a work-around. But I'd like a new ticket to know what is happening for LSM because off-hand it should work for LSM. We should have a bug-fixing ticket to investigate that.

Thanks, @sueloverso for pointing this out. I created a ticket WT-7090 last week to fix these failures.

sueloverso reviewed (approved):
These changes LGTM. I have one request. Can you itemize, in WT-7090, the locations of all the things that need to be updated when that is fixed so that none of these places is forgotten (like the documentation).

kommiharibabu reviewed (approved):
LGTM
